python/nwm_client/src/hydrotools/nwm_client/UnitHandler.py

```diff
@@ -0,0 +1,82 @@
+import pint
+from dataclasses import dataclass
+import numpy.typing as npt
+import pandas as pd
+import sys
+
+def _default_registry() -> pint.UnitRegistry:
+    """Get a default registry depending upon Python version."""
+    version = sys.version_info.major + sys.version_info.minor/10.0
+    if version < 3.8:
+        return pint.UnitRegistry()
+    return pint.UnitRegistry(cache_folder=":auto:")
+
+@dataclass
+class UnitHandler:
+    """Engine to handle unit of measurement conversions.
+
+    Attributes
+    ----------
+    unit_registry: pint.UnitRegistry, default pint.UnitRegistry(cache_folder=":auto:")
+        pint.UnitRegistry that handles all units used by UnitHandler.
+    
+    """
+    unit_registry: pint.UnitRegistry = _default_registry()
+
+    def conversion_factor(self, from_units: str, to_units: str) -> float:
+        """Compute and return a conversion factor from from_units to 
+        to_units.
+        
+        Parameters
+        ----------
+        from_units: pint.UnitRegistry.Quantity compatible str
+            Units from which to convert (e.g. "m^3/s")
+        to_units: pint.UnitRegistry.Quantity compatible str
+            Desired conversion units (e.g. "ft^3/s")
+            
+        Returns
+        -------
+        result: float
+            Conversion factor to transform from_units to to_units.
+
+        Example
+        -------
+        >>> unit_handler = UnitHandler.UnitHandler()
+        >>> unit_handler.conversion_factor("ft", "m")
+        0.30479999999999996
+        """
+        # Return conversion factor
+        return self.unit_registry.Quantity(1, from_units).to(to_units).magnitude
+
+    def convert_values(self, value: npt.ArrayLike, from_units: str, to_units: str) -> npt.ArrayLike:
+        """Convert value from from_units to to_units.
+        
+        Parameters
+        ----------
+        value: array-like, required
+            Values to convert.
+        from_units: pint.UnitRegistry.Quantity compatible str
+            Units from which to convert (e.g. "m^3/s")
+        to_units: pint.UnitRegistry.Quantity compatible str
+            Desired conversion units (e.g. "ft^3/s")
+            
+        Returns
+        -------
+        result: array-like
+            Converted values same shape as value.
+
+        Example
+        -------
+        >>> unit_handler = UnitHandler.UnitHandler()
+        >>> unit_handler.convert_values([1, 1, 1], "ft", "m")
+        array([0.3048, 0.3048, 0.3048])
+        """
+        # Check for pandas.series
+        if isinstance(value, pd.Series):
```

Comment:
Does using an arraylike view here work?  If so, it would simplify this code and remove the `isinstance` check I think.
```python
converted = self.unit_registry.Quanitity(value[:], from_units).to(to_units).magnitude
value[:] = converted
return value
```
This would essentially do an inplace update and return the reference to the thing updated.  If you really wanted a copy, then the existing might be the simplest?

Reply:
This works about the same as I was thinking, but does have a caveat of only really being applicable to values, because of the slice/view here:
```python
df.loc[:, "value"] = self._unit_handler.convert_values(df["value"], "m^3/s", "ft^3/s")
```
the values get updated, but something like this wouldn't work (which is what I thought you were looking for)
```python
s = pd.Series([1,1,1])
s_converted = convert_values(s, "m^3/s", "ft^3/s")
```
`s_converted` would be an array, not a series.
I wonder if it would be worth writing this in a way you could `apply` or `map` the function?
```python
s_converted = s.apply(convert_values, args=("m^3/s", "ft^3/s"))
```

Though I think what you have here is both useful and efficient for the time being.